### PR TITLE
[PAYARA-3860] - Added synchronization block double check to avoid race condition

### PR DESCRIPTION
**Pull request by cubastanley**
Added a synchronized double check in PayaraInstanceImpl to avoid race condition - tested through sheer brute forcing of Thread.sleep() and spamming of functionality and problem no longer appears to be present

### CODE REVIEW

**Comment by cubastanley** [comment]
Jenkins test please

File: appserver/payara-appserver-modules/payara-micro-service/src/main/java/fish/payara/appserver/micro/services/PayaraInstanceImpl.java [hunk 1]
```diff
@@ -243,25 +240,29 @@ public void event(Event event) {
...
-        } 
-        // Adds the application to the clustered register of deployed applications
+        } // Adds the application to the clustered register of deployed applications
         else if (event.is(Deployment.APPLICATION_STARTED)) {
             if (event.hook() != null && event.hook() instanceof ApplicationInfo) {
                 ApplicationInfo applicationInfo = (ApplicationInfo) event.hook();
-                if (me == null) {//race check
-                    initialiseInstanceDescriptor();
+                //Race condition double check
+                if (me == null) {
```
Comment (Pandrex247) [inline appserver/payara-appserver-modules/payara-micro-service/src/main/java/fish/payara/appserver/micro/services/PayaraInstanceImpl.java]: https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_Java
https://www.javaworld.com/article/2074979/double-checked-locking--clever--but-broken.html
Page 348 of the book I'm about to plonk on your desk :)

Comment (pdudits) [reply appserver/payara-appserver-modules/payara-micro-service/src/main/java/fish/payara/appserver/micro/services/PayaraInstanceImpl.java]: Oh, why did I miss that article? :)

Likely the easiest solution then is to replace any access to `me` with synchronized getter that can also initialize the descriptor when null.

Comment (cubastanley) [reply appserver/payara-appserver-modules/payara-micro-service/src/main/java/fish/payara/appserver/micro/services/PayaraInstanceImpl.java]: Is something like
`private synchronized InstanceDescriptorImpl getMyInstanceDescriptor() {
        if(me == null) {
            initialiseInstanceDescriptor();
        }
        return me;
    }`
the kind of thing you mean?

**Comment by cubastanley** [comment]
Jenkins test please

File: appserver/payara-appserver-modules/payara-micro-service/src/main/java/fish/payara/appserver/micro/services/PayaraInstanceImpl.java [hunk 2]
```diff
@@ -139,19 +138,26 @@
...
 
     @Inject
     private ApplicationRegistry appRegistry;
-    
+
     @Inject
     private PayaraExecutorService executor;
     
+    private synchronized InstanceDescriptorImpl getMyInstanceDescriptor() {
+        if(me == null) {
```
Comment (pdudits) [inline appserver/payara-appserver-modules/payara-micro-service/src/main/java/fish/payara/appserver/micro/services/PayaraInstanceImpl.java]: space after if, please

File: appserver/payara-appserver-modules/payara-micro-service/src/main/java/fish/payara/appserver/micro/services/PayaraInstanceImpl.java [hunk 3]
```diff
@@ -239,50 +243,46 @@ void postConstruct() {
     @SuppressWarnings({"unchecked"})
     public void event(Event event) {
         if (event.is(EventTypes.SERVER_READY)) {
-            initialiseInstanceDescriptor();
-            PayaraInternalEvent pie = new PayaraInternalEvent(PayaraInternalEvent.MESSAGE.ADDED, me);
+            //initialiseInstanceDescriptor();
```
Comment (pdudits) [inline appserver/payara-appserver-modules/payara-micro-service/src/main/java/fish/payara/appserver/micro/services/PayaraInstanceImpl.java]: remove this one as well

**Comment by cubastanley** [comment]
Jenkins Test Please

**Pull request merged**